materialize-bigquery/transactor.go

```diff
@@ -2,287 +2,271 @@ package main
-	var ctx = it.Context()
-
-	// Iterate through all of the Load requests being made.
-	for it.Next() {
+func RunTransactor(ctx context.Context, cfg *config, stream pm.Driver_TransactionsServer, open *pm.TransactionRequest_Open) error {
+	bigqueryClient, err := cfg.BigQueryClient(ctx)
+	if err != nil {
+		return err
+	}
 
-		var b = t.bindings[it.Binding]
+	storageClient, err := cfg.StorageClient(ctx)
+	if err != nil {
+		return err
+	}
 
-		// Check to see if we have a keyFile open already in gcs for this binding and if not, create one.
-		if b.load.keysFile == nil {
-			b.load.keysFile, err = t.ep.NewExternalDataConnectionFile(
-				ctx,
-				tmpFileName(),
-				b.load.extDataConfig,
-			)
-			if err != nil {
-				return fmt.Errorf("new external data connection file: %v", err)
-			}
+	t := &transactor{
+		config:          cfg,
+		bigqueryClient:  bigqueryClient,
+		storageClient:   storageClient,
+		materialization: open.Materialization,
+		checkpoint:      NewBigQueryCheckPoint(),
+		bindings:        make([]*Binding, len(open.Materialization.Bindings)),
+	}
 
-			// Clean up the temporary files when load complete (or we error out).
-			defer func() {
-				if err := b.load.keysFile.Delete(ctx); err != nil {
-					log.Errorf("could not delete load keyfile: %v", err)
-				}
-				b.load.keysFile = nil // blank it
-			}()
+	if open.DriverCheckpointJson != nil {
+		if err := json.Unmarshal(open.DriverCheckpointJson, &t.checkpoint); err != nil {
+			return fmt.Errorf("parsing driver config: %w", err)
 		}
+	}
 
-		// Convert our key tuple into a slice of values appropriate for the database.
-		convertedKey, err := b.load.paramsConverter.Convert(it.Key)
+	for i, sb := range open.Materialization.Bindings {
+		binding, err := NewBinding(ctx, cfg, storageClient.Bucket(cfg.Bucket), sb)
 		if err != nil {
-			return fmt.Errorf("converting key: %w", err)
+			return fmt.Errorf("binding generation: %w", err)
 		}
 
-		// Write our row of keys to the keyFile.
-		if err = b.load.keysFile.WriteRow(convertedKey); err != nil {
-			return fmt.Errorf("writing normalized key to keyfile: %w", err)
-		}
+		t.bindings[i] = binding
 
-	}
-	if it.Err() != nil {
-		return it.Err()
-	}
-
-	// Build the queries of all documents across all bindings that were requested.
-	var subqueries []string
-	// This is the map of external table references we will populate.
-	var edcTableDefs = make(map[string]bigquery.ExternalData)
-	for _, b := range t.bindings {
-		// If we have a data file for this round.
-		if b.load.keysFile != nil {
-
-			// Flush and close the keyfile.
-			if err := b.load.keysFile.Close(); err != nil {
-				return fmt.Errorf("closing external data connection file: %w", err)
+		// Error here doesn't mean it's fatal, it just means the checkpoint didn't have
+		// a checkpoint binding, which is a normal scenario if the last time the driver
+		// stored a checkpoint, the binding didn't exist.
+		if driverBinding, err := t.checkpoint.Binding(i); err == nil {
```

Comment:
This seems a bit odd to me: I think you are somehow mixing the new bindings with the old bindings from the checkpoint, but I don't think they are mixable at all.

You might have more or less number of bindings and their ordering is not going to be the same, so you can't correspond them with each other in any way.

Consider this example:
MaterializationSpec v1 creates checkpoint for bindings:
1. A
2. B
3. C
4. D

MaterializationSpec v2 sends an Open request to the connector, and has these bindings:
1. A
2. G
3. D

They can't be mixed with each other at all. The new Materialization might replace all previous bindings with new ones that are not related.

In Acknowledge, you cannot rely on `t.bindings`, and all your operations must solely depend on the checkpoint, otherwise the potentially new MaterializationSpec that comes with a new transaction can still interfere with the old MaterializationSpec for which we have a checkpoint, but we should protect against that.

Reply:
I think this new version fixes this.